File: apps/server/src/services/project-service.ts
import { randomUUID } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { homedir } from "node:os";
import type { Project, ProjectInput, ProjectReorderInput, Service, ServiceInput } from "@devdeck/shared";

export class ProjectNotFoundError extends Error {
  constructor(projectId: string) {
    super(`Project ${projectId} was not found.`);
    this.name = "ProjectNotFoundError";
  }
}

export class ProjectValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProjectValidationError";
  }
}

function defaultDataDirectory() {
  return process.env.DEVDECK_DATA_DIR ?? join(process.env.APPDATA ?? homedir(), "DevDeck");
}

function normalizeOptionalString(value: string | undefined) {
  const normalized = value?.trim();
  return normalized ? normalized : undefined;
}

function normalizeService(input: ServiceInput, existingId?: string): Service {
  if (!input || typeof input !== "object") {
    throw new ProjectValidationError("Every service must be an object.");
  }

  const name = typeof input.name === "string" ? input.name.trim() : "";
  const command = typeof input.command === "string" ? input.command.trim() : "";

  if (!name) {
    throw new ProjectValidationError("Every service needs a name.");
  }

  if (!command) {
    throw new ProjectValidationError(`Service ${name} needs a command.`);
  }

  if (input.port !== undefined && (!Number.isInteger(input.port) || input.port < 1 || input.port > 65535)) {
    throw new ProjectValidationError(`Service ${name} has an invalid port.`);
  }

  return {
    id: existingId ?? input.id ?? randomUUID(),
    name,
    command,
    cwd: normalizeOptionalString(input.cwd),
    port: input.port
  };
}

function normalizeProject(input: ProjectInput, existing?: Project): Project {
  if (!input || typeof input !== "object") {
    throw new ProjectValidationError("Project data must be an object.");
  }

  const name = typeof input.name === "string" ? input.name.trim() : "";
  const projectPath = typeof input.path === "string" ? input.path.trim() : "";

  if (!name) {
    throw new ProjectValidationError("A project needs a name.");
  }

  if (!projectPath) {
    throw new ProjectValidationError("A project needs a local path.");
  }

  if (input.services !== undefined && !Array.isArray(input.services)) {
    throw new ProjectValidationError("Project services must be an array.");
  }

  const incomingServices = input.services ?? [];
  const existingServices = new Map((existing?.services ?? []).map((service) => [service.id, service]));

  return {
    id: existing?.id ?? randomUUID(),
    name,
    path: projectPath,
    services: incomingServices.map((service) => normalizeService(service, service.id ? existingServices.get(service.id)?.id : undefined))
  };
}

export class ProjectService {
  private readonly filePath: string;
  private projects: Project[] | null = null;

  constructor(filePath = join(defaultDataDirectory(), "projects.json")) {
    this.filePath = filePath;
  }

  async list() {
    await this.ensureLoaded();
    return this.projects ?? [];
  }

  async get(projectId: string) {
    const projects = await this.list();
    const project = projects.find((candidate) => candidate.id === projectId);

    if (!project) {
      throw new ProjectNotFoundError(projectId);
    }

    return project;
  }

  async create(input: ProjectInput) {
    const projects = await this.list();
    const project = normalizeProject(input);

    projects.push(project);
    await this.persist();
    return project;
  }

  async update(projectId: string, input: ProjectInput) {
    const projects = await this.list();
    const projectIndex = projects.findIndex((candidate) => candidate.id === projectId);

    if (projectIndex === -1) {
      throw new ProjectNotFoundError(projectId);
    }

    const project = normalizeProject(input, projects[projectIndex]);
    projects[projectIndex] = project;
    await this.persist();
    return project;
  }

  async remove(projectId: string) {
    const projects = await this.list();
    const projectIndex = projects.findIndex((candidate) => candidate.id === projectId);

    if (projectIndex === -1) {
      throw new ProjectNotFoundError(projectId);
    }

    projects.splice(projectIndex, 1);
    await this.persist();
  }

  async reorder(input: ProjectReorderInput) {
    const projects = await this.list();

    if (!Array.isArray(input.projectIds) || input.projectIds.length !== projects.length) {
      throw new ProjectValidationError("Project reorder must include every registered project exactly once.");
    }

    const registeredIds = new Set(projects.map((project) => project.id));
    const requestedIds = new Set(input.projectIds);
    if (requestedIds.size !== input.projectIds.length || input.projectIds.some((projectId) => !registeredIds.has(projectId))) {
      throw new ProjectValidationError("Project reorder contains an unknown or duplicate project.");
    }

    this.projects = input.projectIds.map((projectId) => projects.find((project) => project.id === projectId)!);
    await this.persist();
    return this.projects;
  }

  private async ensureLoaded() {
    if (this.projects) {
      return;
    }

    try {
      const contents = await readFile(this.filePath, "utf8");
      const parsed = JSON.parse(contents) as unknown;

      if (!Array.isArray(parsed)) {
        throw new Error("Project data must be an array.");
      }

      this.projects = parsed as Project[];
    } catch (error) {
      if (isFileNotFound(error)) {
        this.projects = [];
        await this.persist();
        return;
      }

      throw new Error(`Unable to read DevDeck project data: ${error instanceof Error ? error.message : "unknown error"}`);
    }
  }

  private async persist() {
    await mkdir(dirname(this.filePath), { recursive: true });
    await writeFile(this.filePath, `${JSON.stringify(this.projects ?? [], null, 2)}\n`, "utf8");
  }
}

function isFileNotFound(error: unknown): error is NodeJS.ErrnoException {
  return typeof error === "object" && error !== null && "code" in error && error.code === "ENOENT";
}

File: apps/desktop/src-tauri/src/services/project_service.rs
use crate::error::{AppError, AppResult};
use crate::models::{Project, ProjectInput, ProjectReorderInput, Service, ServiceInput};
use crate::state::AppState;
use std::path::Path;
use uuid::Uuid;

pub fn get_projects(state: &AppState) -> AppResult<Vec<Project>> {
    let guard = state
        .inner
        .lock()
        .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
    Ok(guard.projects.clone())
}

pub fn get_project(state: &AppState, project_id: &str) -> AppResult<Project> {
    let guard = state
        .inner
        .lock()
        .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
    guard
        .projects
        .iter()
        .find(|project| project.id == project_id)
        .cloned()
        .ok_or_else(|| AppError::ProjectNotFound(project_id.into()))
}

pub fn add_project(state: &AppState, input: ProjectInput) -> AppResult<Project> {
    let project = normalize_project(input, None)?;
    {
        let mut guard = state
            .inner
            .lock()
            .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
        guard.projects.push(project.clone());
    }
    state.persist()?;
    Ok(project)
}

pub fn update_project(
    state: &AppState,
    project_id: &str,
    input: ProjectInput,
) -> AppResult<Project> {
    let existing = get_project(state, project_id)?;
    let project = normalize_project(input, Some(&existing))?;
    {
        let mut guard = state
            .inner
            .lock()
            .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
        let index = guard
            .projects
            .iter()
            .position(|candidate| candidate.id == project_id)
            .ok_or_else(|| AppError::ProjectNotFound(project_id.into()))?;
        guard.projects[index] = project.clone();
    }
    state.persist()?;
    Ok(project)
}

pub fn remove_project(state: &AppState, project_id: &str) -> AppResult<()> {
    let mut guard = state
        .inner
        .lock()
        .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
    let index = guard
        .projects
        .iter()
        .position(|project| project.id == project_id)
        .ok_or_else(|| AppError::ProjectNotFound(project_id.into()))?;
    guard.projects.remove(index);
    drop(guard);
    state.persist()
}

pub fn reorder_projects(state: &AppState, input: ProjectReorderInput) -> AppResult<Vec<Project>> {
    let mut guard = state
        .inner
        .lock()
        .map_err(|_| AppError::Storage("State lock poisoned.".into()))?;
    if input.project_ids.len() != guard.projects.len() {
        return Err(AppError::Storage(
            "Project reorder must include every project exactly once.".into(),
        ));
    }

    let mut reordered = Vec::with_capacity(guard.projects.len());
    for project_id in input.project_ids {
        let project = guard
            .projects
            .iter()
            .find(|project| project.id == project_id)
            .cloned()
            .ok_or_else(|| AppError::ProjectNotFound(project_id.clone()))?;
        if reordered
            .iter()
            .any(|candidate: &Project| candidate.id == project.id)
        {
            return Err(AppError::Storage(
                "Project reorder contains a duplicate project.".into(),
            ));
        }
        reordered.push(project);
    }

    guard.projects = reordered.clone();
    drop(guard);
    state.persist()?;
    Ok(reordered)
}

fn normalize_project(input: ProjectInput, existing: Option<&Project>) -> AppResult<Project> {
    let name = input.name.trim().to_string();
    let path = input.path.trim().to_string();
    if name.is_empty() {
        return Err(AppError::InvalidProjectName);
    }
    if path.is_empty() {
        return Err(AppError::InvalidProjectPath(
            "A project path is required.".into(),
        ));
    }
    if !Path::new(&path).is_dir() {
        return Err(AppError::InvalidProjectPath(format!(
            "Project directory does not exist: {path}"
        )));
    }

    let existing_services = existing.map(|project| &project.services);
    let services = input
        .services
        .unwrap_or_default()
        .into_iter()
        .map(|service| normalize_service(service, existing_services))
        .collect::<AppResult<Vec<_>>>()?;
    Ok(Project {
        id: existing
            .map(|project| project.id.clone())
            .unwrap_or_else(|| Uuid::new_v4().to_string()),
        name,
        path,
        services,
    })
}

fn normalize_service(input: ServiceInput, existing: Option<&Vec<Service>>) -> AppResult<Service> {
    let name = input.name.trim().to_string();
    let command = input.command.trim().to_string();
    if name.is_empty() || command.is_empty() {
        return Err(AppError::InvalidService);
    }
    let id = input
        .id
        .filter(|id| {
            existing
                .map(|services| services.iter().any(|service| service.id == *id))
                .unwrap_or(false)
        })
        .unwrap_or_else(|| Uuid::new_v4().to_string());
    Ok(Service {
        id,
        name,
        command,
        cwd: input
            .cwd
            .and_then(|value| (!value.trim().is_empty()).then(|| value.trim().to_string())),
        port: input.port,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::AppState;
    use std::fs;

    #[test]
    fn project_crud_and_order_are_persisted() {
        let test_dir = std::env::temp_dir().join(format!("devdeck-test-{}", Uuid::new_v4()));
        fs::create_dir_all(&test_dir).expect("create test directory");
        let state = AppState::load(test_dir.join("projects.json")).expect("load state");

        let first = add_project(
            &state,
            ProjectInput {
                name: "First".into(),
                path: test_dir.to_string_lossy().into_owned(),
                services: None,
            },
        )
        .expect("add first project");
        let second = add_project(
            &state,
            ProjectInput {
                name: "Second".into(),
                path: test_dir.to_string_lossy().into_owned(),
                services: None,
            },
        )
        .expect("add second project");

        let reordered = reorder_projects(
            &state,
            ProjectReorderInput {
                project_ids: vec![second.id.clone(), first.id.clone()],
            },
        )
        .expect("reorder projects");
        assert_eq!(reordered[0].id, second.id);

        let loaded = AppState::load(test_dir.join("projects.json")).expect("reload state");
        assert_eq!(
            get_projects(&loaded).expect("read projects")[0].id,
            second.id
        );

        remove_project(&loaded, &first.id).expect("remove first project");
        remove_project(&loaded, &second.id).expect("remove second project");
        fs::remove_dir_all(test_dir).expect("remove test directory");
    }

    #[test]
    fn rejects_missing_project_directory() {
        let state_file = std::env::temp_dir().join(format!("devdeck-test-{}.json", Uuid::new_v4()));
        let state = AppState::load(state_file.clone()).expect("load state");
        let result = add_project(
            &state,
            ProjectInput {
                name: "Missing".into(),
                path: state_file
                    .with_extension("missing")
                    .to_string_lossy()
                    .into_owned(),
                services: None,
            },
        );
        assert!(matches!(result, Err(AppError::InvalidProjectPath(_))));
        let _ = fs::remove_file(state_file);
    }
}
